<a href="https://colab.research.google.com/github/anandababugudipudi/VADER-Sentiment-Analysis/blob/main/VADER_Sentiment_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**VADER Sentiment Analysis**
**VADER (Valence Aware Dictionary and sEntiment Reasoner)** is a lexicon and rule-based sentiment analysis tool that is specifically attuned to sentiments expressed in social media. VADER uses a combination of A sentiment lexicon is a list of lexical features (e.g., words) which are generally labelled according to their semantic orientation as either positive or negative.

VADER has been found to be quite successful when dealing with social media texts, NY Times editorials, movie reviews, and product reviews. This is because VADER not only tells about the Positivity and Negativity score but also tells us about how positive or negative a sentiment is.

It is fully open-sourced under the MIT License. The developers of VADER have used Amazon’s Mechanical Turk to get most of their ratings, You can find complete details on their [Github Page](https://github.com/cjhutto/vaderSentiment).

###**Advantages of using VADER**
VADER has a lot of advantages over traditional methods of Sentiment Analysis, including:
- It works exceedingly well on social media type text, yet readily generalizes to multiple domains
- It doesn’t require any training data but is constructed from a generalizable, valence-based, human-curated gold standard sentiment lexicon
- It is fast enough to be used online with streaming data, and
- It does not severely suffer from a speed-performance tradeoff.

The source of this article is a very easy to read paper published by the creaters of VADER library.You can read the paper [here](http://comp.social.gatech.edu/papers/icwsm14.vader.hutto.pdf).

###**What is Sentiment Analysis?**
Sentiment Analysis, or Opinion Mining, is a sub-field of Natural Language Processing (NLP) that tries to identify and extract opinions within a given text. The aim of sentiment analysis is to gauge the attitude, sentiments, evaluations, attitudes and emotions of a speaker/writer based on the computational treatment of subjectivity in a text.

###**Installation**
The simplest way is to use the command line to do an installation from [PyPI](https://pypi.python.org/pypi/vaderSentiment) using pip. Check their [Github repository](https://github.com/cjhutto/vaderSentiment) for the detailed explanation.

In [1]:
!pip install vaderSentiment

     |████████████████████████████████| 133kB 8.8MB/s 


###**Import the Package**

In [2]:
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

Create an analyser instance using `SentimentIntensityAnalyzer` object

In [3]:
analyser = SentimentIntensityAnalyzer()

###**Working & Scoring**
Let us test our first sentiment using VADER now. We will use the polarity_scores() method to obtain the polarity indices for the given sentence.

In [4]:
def sentiment_analyzer_scores(sentence):
    score = analyser.polarity_scores(sentence)
    print("{:-<40} {}".format(sentence, str(score)))

Let us check how VADER performs on a given review:

In [5]:
sentiment_analyzer_scores("The phone is super cool.")

The phone is super cool.---------------- {'neg': 0.0, 'neu': 0.326, 'pos': 0.674, 'compound': 0.7351}


- The Positive, Negative and Neutral scores represent the proportion of text that falls in these categories. This means our sentence was rated as 67% Positive, 33% Neutral and 0% Negative. Hence all these should add up to 1.
- The Compound score is a metric that calculates the sum of all the lexicon ratings which have been normalized between -1(most extreme negative) and +1 (most extreme positive). In the case above, lexicon ratings for andsupercool are 2.9and respectively1.3. The compound score turns out to be 0.75 , denoting a very high positive sentiment.

Read [here](https://github.com/cjhutto/vaderSentiment#about-the-scoring) for more details on VADER scoring methodology.

**VADER analyses sentiments primarily based on certain key points:**
- **Punctuation:** The use of an exclamation mark(!), increases the magnitude of the intensity without modifying the semantic orientation. For example, “The food here is good!” is more intense than “The food here is good.” and an increase in the number of (!), increases the magnitude accordingly.


In [6]:
# Baselien Sentence
sentiment_analyzer_scores("The food here is good.")

The food here is good.------------------ {'neg': 0.0, 'neu': 0.58, 'pos': 0.42, 'compound': 0.4404}


In [7]:
# Punctuation
sentiment_analyzer_scores("The food here is good!")
sentiment_analyzer_scores("The food here is good!!")
sentiment_analyzer_scores("The food here is good!!!")

The food here is good!------------------ {'neg': 0.0, 'neu': 0.556, 'pos': 0.444, 'compound': 0.4926}
The food here is good!!----------------- {'neg': 0.0, 'neu': 0.534, 'pos': 0.466, 'compound': 0.5399}
The food here is good!!!---------------- {'neg': 0.0, 'neu': 0.514, 'pos': 0.486, 'compound': 0.5826}


**Re:** We can observe that the overall compound score is increasing with the increase in exclamation marks.

- **Capitalization:** Using upper case letters to emphasize a sentiment-relevant word in the presence of other non-capitalized words, increases the magnitude of the sentiment intensity. For example, “The food here is GREAT!” conveys more intensity than “The food here is great!”

In [8]:
# Baselien Sentence
sentiment_analyzer_scores("The food here is great!")

The food here is great!----------------- {'neg': 0.0, 'neu': 0.477, 'pos': 0.523, 'compound': 0.6588}


In [9]:
# Capitalization
# Baselien Sentence
sentiment_analyzer_scores("The food here is GREAT!")

The food here is GREAT!----------------- {'neg': 0.0, 'neu': 0.438, 'pos': 0.562, 'compound': 0.729}


**Re:** Here we can observe that by changing the letters to capital the compound score has increased to 0.729 from 0.6588.

- **Degree modifiers:** Also called intensifiers, they impact the sentiment intensity by either increasing or decreasing the intensity. For example, “The service here is extremely good” is more intense than “The service here is good”, whereas “The service here is marginally good” reduces the intensity.


In [10]:
# Baseline Sentence
sentiment_analyzer_scores("The service here is good")

The service here is good---------------- {'neg': 0.0, 'neu': 0.58, 'pos': 0.42, 'compound': 0.4404}


In [11]:
# Degree Modifiers
sentiment_analyzer_scores("The service here is extremenly good")
sentiment_analyzer_scores("The service here is marginally good")

The service here is extremenly good----- {'neg': 0.0, 'neu': 0.633, 'pos': 0.367, 'compound': 0.4404}
The service here is marginally good----- {'neg': 0.0, 'neu': 0.657, 'pos': 0.343, 'compound': 0.3832}


**Re:** The score has been reduced when we have used the word *marginally*.

- **Conjunctions:** Use of conjunctions like “but” signals a shift in sentiment polarity, with the sentiment of the text following the conjunction being dominant. “The food here is great, but the service is horrible” has mixed sentiment, with the latter half dictating the overall rating.

In [12]:
# Baseline Sentence
sentiment_analyzer_scores("The food here is great. The service is horrible.")

The food here is great. The service is horrible. {'neg': 0.24, 'neu': 0.479, 'pos': 0.281, 'compound': 0.1531}


In [13]:
# Conjunctions
sentiment_analyzer_scores("The food here is great, but the service is horrible.")

The food here is great, but the service is horrible. {'neg': 0.31, 'neu': 0.523, 'pos': 0.167, 'compound': -0.4939}


**Re:** The score has been reduced when we have used the word *but*.

- **Preceding Tri-gram:** By examining the tri-gram preceding a sentiment-laden lexical feature, we catch nearly 90% of cases where negation flips the polarity of the text. A negated sentence would be “The food here isn’t really all that great”.

###**Handling Emojis, Slangs, and Emoticons:**
VADER performs very well with emojis, slangs, and acronyms in sentences. Let us see each with an example.

- **Emojis:**

In [17]:
sentiment_analyzer_scores('I am 😄 today')
sentiment_analyzer_scores('😊')
sentiment_analyzer_scores('😥')
sentiment_analyzer_scores('☹️')
sentiment_analyzer_scores('💘')

I am 😄 today---------------------------- {'neg': 0.0, 'neu': 0.522, 'pos': 0.478, 'compound': 0.6705}
😊--------------------------------------- {'neg': 0.0, 'neu': 0.333, 'pos': 0.667, 'compound': 0.7184}
😥--------------------------------------- {'neg': 0.275, 'neu': 0.268, 'pos': 0.456, 'compound': 0.3291}
☹️-------------------------------------- {'neg': 0.706, 'neu': 0.294, 'pos': 0.0, 'compound': -0.34}
💘--------------------------------------- {'neg': 0.0, 'neu': 0.323, 'pos': 0.677, 'compound': 0.6369}


- **Slangs:**

In [18]:
sentiment_analyzer_scores("Today SUX!")
sentiment_analyzer_scores("Today only kinda sux! But I'll get by, lol")

Today SUX!------------------------------ {'neg': 0.779, 'neu': 0.221, 'pos': 0.0, 'compound': -0.5461}
Today only kinda sux! But I'll get by, lol {'neg': 0.127, 'neu': 0.556, 'pos': 0.317, 'compound': 0.5249}


- **Emoticons:**

In [19]:
sentiment_analyzer_scores("Make sure you :) or :D today!")

Make sure you :) or :D today!----------- {'neg': 0.0, 'neu': 0.294, 'pos': 0.706, 'compound': 0.8633}


###**Conclusion:**
The results of VADER analysis are not only remarkable but also very encouraging. The outcomes highlight the tremendous benefits that can be attained by the use of VADER in cases of micro-blogging sites wherein the text data is a complex mix of a variety of text.